### PR TITLE
Add a sophisticated "maintenance" demo

### PR DESCRIPTION
**Pull request by timholy**
I'm not sure what folks think about having something this "sophisticated" (aka, likely to go stale...), but at a time when many people need to be transitioning from travis to GitHub actions, this script can take most of the manual work away.

It relies on https://github.com/JuliaWeb/GitHub.jl/pull/179, so we shouldn't merge this until its fate is decided.

### CODE REVIEW

File: demos/maintenance.jl [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+# This demo adds or updates workflow files (TagBot, Documenter, and CI) for a set of repositories
+# in an organization. In general you should copy/paste these lines after editing them
+# to suit your needs. Note that you'll need to specify your GitHub personal access token,
+# specify the organization/user, and select repositories interactively.
+
+# authenticate
+auth = authenticate(#= your-github-personal-access-token =#)   # don't ever post this publicly!
+
+# Get all the repositories in the JuliaDebug organization
+rs = repos(#= organization/user goes here =#, true; auth)[1]   # change true -> false for users (true = "org")
```
Comment (mschauer) [inline demos/maintenance.jl:13]:
```suggestion
rs = repos(#= organization/user goes here =#, true; auth=auth)[1]   # change true -> false for users (true = "org")
```

Comment (timholy) [reply demos/maintenance.jl:13]: old julia versions are so incapable...

**Review by DilumAluthge — APPROVED** [review]

File: demos/maintenance.jl [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+rs = rs[collect(sel)]
+
+## Key/secret creation
+
+# Create a `DOCUMENTER_KEY` secret and corresponding `Documenter` deploy key, unless the repository
+# already has a secret of that name.
+# This key is used by TagBot as well as Documenter, so do this even if you don't have any `docs/` in your repository.
+for r in rs
+    has_secret = false
+    for s in secrets(r; auth)[1]
```
Comment (mschauer) [inline demos/maintenance.jl:31]:
```suggestion
    for s in secrets(r; auth=auth)[1]
```

File: demos/maintenance.jl [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        if s.name == "DOCUMENTER_KEY"
+            has_secret = true
+            break
+        end
+    end
+    has_secret && continue
+    pubkey, privkey = GitHub.genkeys()
+    create_deploykey(r; auth, params=Dict("key"=>pubkey, "title"=>"Documenter", "read_only"=>false))
```
Comment (mschauer) [inline demos/maintenance.jl:39]:
```suggestion
    create_deploykey(r; auth=auth, params=Dict("key"=>pubkey, "title"=>"Documenter", "read_only"=>false))
```

File: demos/maintenance.jl [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    create_secret(r, "DOCUMENTER_KEY"; auth, value=privkey)
```
Comment (mschauer) [inline demos/maintenance.jl:40]:
```suggestion
    create_secret(r, "DOCUMENTER_KEY"; auth=auth, value=privkey)
```

File: demos/maintenance.jl [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+
+## TagBot
+
+# Add or modify the TagBot workflow to use `DOCUMENTER_KEY`. Here we use the templates in MassInstallAction directly,
+# but alternatively we could create a workflow script here and use `Workflow("TagBot", "TagBot.yml" => script)`.
+#
+# Related: https://discourse.julialang.org/t/ann-required-updates-to-tagbot-yml/49249
+workflow = MassInstallAction.tag_bot()
+for r in rs
+    MassInstallAction.install(workflow, r; auth, commit_message="Set up TagBot workflow")
```
Comment (mschauer) [inline demos/maintenance.jl:51]:
```suggestion
    MassInstallAction.install(workflow, r; auth=auth, commit_message="Set up TagBot workflow")
```

File: demos/maintenance.jl [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+        - uses: julia-actions/julia-buildpkg@latest
+        - uses: julia-actions/julia-docdeploy@latest
+        env:
+            GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
+            DOCUMENTER_KEY: \${{ secrets.DOCUMENTER_KEY }}
+"""
+workflow = MassInstallAction.Workflow("Documenter", "Documenter.yml" => script)
+for r in rs
+    # Check whether the repo has docs, if not skip it
+    filesdirs = directory(r, "."; auth)[1]
```
Comment (mschauer) [inline demos/maintenance.jl:79]:
```suggestion
    filesdirs = directory(r, "."; auth=auth)[1]
```

File: demos/maintenance.jl [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    idx = findfirst(fd -> fd.typ == "dir" && fd.name == "docs", filesdirs)
+    if idx !== nothing
+        MassInstallAction.install(workflow, r; auth, commit_message="Build docs on GitHub Actions")
```
Comment (mschauer) [inline demos/maintenance.jl:82]:
```suggestion
        MassInstallAction.install(workflow, r; auth=auth, commit_message="Build docs on GitHub Actions")
```

File: demos/maintenance.jl [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+      - uses: julia-actions/julia-runtest@v1
+      - uses: julia-actions/julia-processcoverage@v1
+      - uses: codecov/codecov-action@v1
+        with:
+          file: lcov.info
+"""
+for r in rs
+    try
+        # Parse the .travis.yml file and extract the versions
+        url = file(r, ".travis.yml"; auth).download_url
```
Comment (mschauer) [inline demos/maintenance.jl:140]:
```suggestion
        url = file(r, ".travis.yml"; auth=auth).download_url
```

File: demos/maintenance.jl [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+            idx = findfirst(fd -> fd.typ == "file" && fd.name ∈ ("appveyor.yml", ".appveyor.yml"), filesdirs)
+            if idx !== nothing
+                push!(todelete, "../../" * filesdirs[idx].name)
+            end
+            workflow = MassInstallAction.Workflow("CI", "CI.yml" => script, todelete)
+            println("\n\n$(r.name) had .travis.yml script:\n", tscript, "\nInstalling the following CI.yml script:\n", script)
+            println("OK to submit pull request? (y/n)")
+            resp = readline(; keep=false)
+            if resp ∈ ("y", "Y", "yes", "Yes", "YES")
+                MassInstallAction.install(workflow, r; auth, commit_message="Switch to GitHub Actions for CI")
```
Comment (mschauer) [inline demos/maintenance.jl:176]:
```suggestion
                MassInstallAction.install(workflow, r; auth=auth, commit_message="Switch to GitHub Actions for CI")
```

File: demos/maintenance.jl [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+                # optional: comment out nightly
+                if v == "nightly"
+                    line = "#" * line
+                end
+                script *= line
+            end
+            script *= ci_post
+            todelete = ["../../.travis.yml"]
+            # Find out if there's an appveyor and set it up for deletion
+            filesdirs = directory(r, "."; auth)[1]
```
Comment (mschauer) [inline demos/maintenance.jl:166]:
```suggestion
            filesdirs = directory(r, "."; auth=auth)[1]
```

**Review by mschauer — COMMENTED** [review]
Did I get all of them?

**Comment by timholy** [comment]
OK, I think this is now ready to go. I turned it into a project, but in the end I opted for `[compat]` bounds rather than checking in the Manifest. Happy to change this if others disagree.

**Pull request merged**